src/lua/protobuf.lua

```diff
@@ -0,0 +1,723 @@
+    local key_count = 0
+    local min_key = 1/0
+    local max_key = -1/0
+    for k, _ in pairs(data) do
+        if type(k) ~= 'number' then
+            error(('Input array for %q repeated field ' ..
+                'contains non-numeric key: %q'):format(field_def.name, k))
+        end
+        if k - math.floor(k) ~= 0 then
+            error(('Input array for %q repeated field contains ' ..
+                'non-integer numeric key: %q'):format(field_def.name, k))
+        end
+        key_count = key_count + 1
+        min_key = math.min(min_key, k)
+        max_key = math.max(max_key, k)
+    end
+    if key_count == 0 then
+        return
+    end
+    if min_key ~= 1 then
+        error(('Input array for %q repeated field got min index %d. ' ..
+            'Must be 1'):format(field_def.name, min_key))
+    end
+    if max_key ~= key_count then
+        error(('Input array for %q repeated field has inconsistent keys. ' ..
+            'Got table with %d fields and max index of %d'):format(
+            field_def.name, key_count, max_key))
+    end
+end
+
+local function validate_type(field_def, value, exp_type)
+    if type(exp_type) == 'table' then
+        local found = false
+        for _, exp_t in pairs(exp_type) do
+            if type(value) == exp_t then
+                found = true
+                break
+            end
+        end
+        if not found then
+            error(('Field %q of %q type gets %q type value.'):format(
+                field_def.name, field_def.type, type(value)))
+        end
+        return
+    end
+    assert(type(exp_type) == 'string')
+    if type(value) ~= exp_type then
+        error(('Field %q of %q type gets %q type value. ' ..
+            'Unsupported or colliding types'):format(field_def.name,
+            field_def.type, type(value)))
+    end
+    return
+end
+
+local function validate_value_error(field_def, value)
+    error(('Input data for %q field is %q and do not fit in %q')
+        :format(field_def.name, value, field_def.type))
+end
+
+local function validate_value(field_def, value, min, max)
+    -- To avoid incorrect comparison unsigned cdata is not
+    -- checked for lower limit
+    if type(value) == 'cdata' and ffi.istype(uint64_t, value) then
+        if value > max then
+            validate_value_error(field_def, value)
+        end
+        return
+    end
+    if value < min or value > max then
+        validate_value_error(field_def, value)
+    end
+end
+
+-- }}} Validations
+
+
+-- {{{ Scalar type definitions
+
+scalars.float = {
+    encode_as_packed = true,
+    validate = function(value, field_def)
+        local MAX_FLOAT = 3.4028234E+38
+        local MIN_FLOAT = -3.4028234E+38
+        validate_type(field_def, value, 'number')
+        validate_value(field_def, math.abs(value), MIN_FLOAT, MAX_FLOAT)
+    end,
+    encode = function(value, field_id)
+        return wireformat.encode_float(value, field_id)
+    end,
+}
+
+scalars.fixed32 = {
```

Comment:
AFAIU, all the `fixed` types represent integral values. However, I see no check that a `number` value to encode has no fractional part. Am I miss something?